notary-commons/src/main/kotlin/sidechain/iroha/ReliableIrohaChainListener.kt

```diff
@@ -58,23 +65,31 @@ class ReliableIrohaChainListener(
      * Returns an observable that emits a new block every time it gets it from Iroha
      */
     override fun getBlockObservable(
-    ): Result<Observable<Pair<iroha.protocol.BlockOuterClass.Block, () -> Unit>>, Exception> {
+    ): Result<Observable<Pair<BlockOuterClass.Block, () -> Unit>>, Exception> {
         return Result.of {
-            val source = PublishSubject.create<Delivery>()
+            val source = PublishSubject.create<Pair<BlockOuterClass.Block, Long>>()
             val deliverCallback = { consumerTag: String, delivery: Delivery ->
                 // This code is executed inside consumerExecutorService
-                source.onNext(delivery)
-            }
-            val obs: Observable<Pair<iroha.protocol.BlockOuterClass.Block, () -> Unit>> = source.map { delivery ->
                 val block = iroha.protocol.BlockOuterClass.Block.parseFrom(delivery.body)
-                logger.info { "New Iroha block from RMQ arrived. Height ${block.blockV1.payload.height}" }
-                Pair(block, {
+                //TODO shall we ignore too old blocks?
+                if (ableToHandleBlock(block)) {
```

Comment:
why do we need such a complicated logic? why can't we just read all the blocks incoming?

Reply:
Because we can consume the same block under some circumstances.